src/v3/src/util/widgetHooks.ts

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { IdxStatus, IdxTransaction } from '@okta/okta-auth-js';
+
+import { HookFunction, HooksOptions, HookType } from '../../../types';
+import { getFormNameForTransaction } from './getEventContext';
+
+/* eslint-disable no-await-in-loop, no-restricted-syntax */
+export async function executeWidgetHooks(
+  hookType: HookType,
+  hooks?: HooksOptions,
+  idxTransaction?: IdxTransaction,
+  error?: Error | null,
+) {
+  if (!hooks || !(idxTransaction || error)) {
+    return;
+  }
+  if (idxTransaction?.status === IdxStatus.SUCCESS) {
+    // Don't execute hooks in the end of authentication flow (OKTA-60410)
+    return;
+  }
+  const formName = idxTransaction ? getFormNameForTransaction(idxTransaction) : 'terminal';
+  if (formName) {
+    const hooksToExecute = hooks[formName]?.[hookType];
+    if (hooksToExecute?.length) {
+      for (const hook of hooksToExecute) {
+        await hook();
+      }
+    }
+  }
+}
+
+export const mergeWidgetHooks = (
+  hooks: HooksOptions, hookType: HookType, formName: string, hook: HookFunction,
+) => {
+  const newHooks = { ...hooks };
+  newHooks[formName] = newHooks[formName] || {};
+  newHooks[formName][hookType] = [
+    ...(newHooks[formName][hookType] || []),
```

Comment:
nice I like this clean way of checking if there are existing hooks and populating them in one line